TLS1.3 support

sam-github opened the pull request:
This introduces TLS1.3 support and makes it the default max protocol,
    but also supports CLI/NODE_OPTIONS switches to disable it if necessary.
    
    TLS1.3 is a major update to the TLS protocol, with many security
    enhancements. It should be preferred over TLS1.2 whenever possible.
    
    TLS1.3 is different enough that even though the OpenSSL APIs are
    technically API/ABI compatible, that when TLS1.3 is negotiated, the
    timing of protocol records and of callbacks broke assumptions hard-coded
    into the 'tls' module.
    
    This change introduces no API incompatibilities when TLS1.2 is
    negotiated. It is the intention that it be backported to current and LTS
    release lines with the default maximum TLS protocol reset to 'TLSv1.2'.
    This will allow users of those lines to explicitly enable TLS1.3 if they
    want.
    
    API incompatibilities between TLS1.2 and TLS1.3 are:
    
    - Variations of `conn.write('data'); conn.destroy()` have undefined
    behaviour according to the streams API. They may or may not send the
    'data', and may or may not cause a ERR_STREAM_DESTROYED error to be
    emitted. This has always been true, but conditions under which the write
    suceeds is slightly but observably different when TLS1.3 is negotiated
    vs when TLS1.2 or below is negotiated.
    
    - If TLS1.3 is negotiated, and a server calls `conn.end()` in its
    'secureConnection' listener without any data being written, the client
    will not receive session tickets (no 'session' events will be emitted,
    and `conn.getSession()` will never return a resumable session).
    
    - The return value of `conn.getSession()` API may not return a resumable
    session if called right after the handshake. The effect will be that
    clients using the legacy `getSession()` API will resume sessions if
    TLS1.2 is negotiated, but will do full handshakes if TLS1.3 is
    negotiated.  See https://github.com/nodejs/node/pull/25831 for more
    information.


##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

sam-github commented:
@bnoordhuis note that I renamed the CLI options you introduced in https://github.com/nodejs/node/pull/23814 that set the min TLS version, because I added options to limit the max, and I wanted whether it was setting the max or min to be clear. The original options haven't landed anywhere because they came with a change to disable TLS1.0 and 1.1, so that rename shouldn't affect anyone, though I'll have to backport part of #23814 with this PR if/when it gets backported

addaleax commented on src/stream_base.h:
```diff
@@ -214,6 +214,13 @@ class StreamResource {
+  //      What happens if Done() is called before return is that onwritecomplete
+  //      gets called sync inside handlWriteReq(), BEFORE afterWriteDispatched()
+  //      is called, but that means it happens before afterWriteDispatched()
+  //      sets req.callback = cb, so the cb never gets called.
```
You're right, this currently only handles async completions. I assume we could do something like we do for `DoShutdown()`, that is, use a return code of `1` for synchronous success?

sam-github replied:
I guess that could work, I guess that positive values aren't meaningful now? They aren't the amount of data written?

addaleax replied:
Yeah, they should currently be unused, we only return 0 or a libuv error code atm.

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/20944/

sam-github commented on test/parallel/test-tls-set-ciphers.js:
```diff
@@ -1,62 +1,93 @@
-// Copyright Joyent, Inc. and other Node contributors.
```
Is this OK? I rewrote this test from scratch, but if wanted, I can put the old copyright back in.

sam-github replied:
@nodejs/TSC or @jasnell (who was involved in restoring incorrectly removed copyrights early in the io.js days, IIRC)

sam-github replied:
@nodejs/tsc

sam-github commented:
@nodejs/crypto PTAL

A note on the test approach: I tried to make as few changes to  the tests as possible, since its supposed to be API almost-compatible. When I had to make significant changes to a test, I often wrapped it, so that the test is run with both TLS1.2 and TLS1.3 as the default (the tests usually negotiate the default max).

Since I want to backport this to release lines where TLS1.2 is the default max, I hacked the default max to be TLSv1.2, and ran `make test` locally. It passes.

In theory, the entire test suite could be run with TLS1.2 and 1.3 (and 1.0, and...), but that seemed excessive. If anyone has concerns about specific tests, I can add coverage for them.

sam-github commented:
I added `lts-watch-10.x`, because I'd like it considered for 10.x (after being in 11.x for sufficient time).

I added `backport-requested-11.x` because I'd like it landed in 11.x, and it can't as-is because of the default.

@nodejs/release @nodejs/lts please adjust if this isn't your expectation.

jasnell commented:
First glance This looks good. Will have to go back over in detail before signing off, however. Thank you for working on this

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/21044/

sam-github commented:
last one didn't even start, try again:

ci: https://ci.nodejs.org/job/node-test-pull-request/21070

jasnell commented on lib/_tls_common.js:
```diff
@@ -148,10 +150,31 @@ exports.createSecureContext = function createSecureContext(options) {
+  // the ciphers and pass them to the appropriate API.
+  const ciphers = (options.ciphers || tls.DEFAULT_CIPHERS).split(':');
+  const cipherList = ciphers.filter((_) => !_.match(/^TLS_/)).join(':');
+  const cipherSuites = ciphers.filter((_) => _.match(/^TLS_/)).join(':');
```
This feels like a complete hack but I'm not sure there's a better way unless we expose two different options for setting the cipher suites.

sam-github replied:
I couldn't think of a better way, and I think two options would be awful. This code may be odd, but its only a couple lines long, and its very robust. We could go the BoringSSL way, and not even allow the TLS1.3 suites to be configurable, but that seems like not such a great idea, either. IoT devs would not be happy.

jasnell commented on src/node_crypto.cc:
```diff
@@ -1256,6 +1316,7 @@ void SecureContext::SetTicketKeys(const FunctionCallbackInfo<Value>& args) {
   ASSIGN_OR_RETURN_UNWRAP(&wrap, args.Holder());
   Environment* env = wrap->env();
 
+  // XXX(sam) Move type and len check to js, and CHECK() in C++.
```
nit: should these be `TODO(@sam-github)` statements instead?

sam-github replied:
OK, I'll change them.

jasnell commented on test/parallel/test-tls-alert-handling.js:
```diff
@@ -40,11 +40,13 @@ const errorHandler = common.mustCall((err) => {
     server.close();
 });
 const server = tls.createServer(opts, common.mustCall(function(s) {
+  console.log('server on secureConnection');
```
nit: would prefer these to be comments rather than console.log statements

sam-github replied:
Its not intended as docs, the docs already say the cb is attached to secureConnection. Debugging the tests is unreasonably hard without logging, every test that failed, the first thing I did was go through and log all the important events. The output isn't very verbose, and it doesn't even show up at all during normal node test runs. There are lots of console.log statements through the test suite (randomly, some have none, some have some). I can take them all out if you all want, but I feel bad for the next person who has to debug these, and will have to add them all back in locally.

jasnell replied:
OK. I still generally don't like extraneous console output in tests but wont block on it.

sam-github replied:
anyone else feel strongly? Its fast to delete them, slow to add them, so I'll do if some more people ask

addaleax replied:
Fwiw, when I add `console.log()` to tests, it’s usually because I think it might be helpful for debugging a future flaky test failure that cannot be reproduced locally. I’m okay with keeping them if you think that that’s the case here.

sam-github replied:
I don't want to side-track this PR with this,  I removed most of the console.log's that I added (a couple remain where they parallel the convention of the pre-existing test).

jasnell commented on test/parallel/test-tls-client-resume.js:
```diff
@@ -44,32 +44,61 @@ const server = tls.Server(options, common.mustCall((socket) => {
+      reconnect();
   }));
 
+  client1.on('data', (d) => {
```
mustCall here?

sam-github replied:
IIRC, whether it happens or not depends on the protocol version. I'll confirm, and document if it differs, and mustCall it if I can.

sam-github replied:
added mustCall

sam-github commented:
Thanks for the review @jasnell, I'll address the comments tomorrow. And the build finished. I see it fails against openssl 1.1.0, I'll go through that.

sam-github commented:
OK, OpenSSL-1.1.0 support has become more difficult than it has in the past.

Compilation is easy, see the last fixup, but what is the expected behaviour? Do we expect the tests to pass?

A node 12.x+openssl-1.1.1 user, when connecting to a TLS1.2or1.3 server will get TLS1.3.

In the same situation, a node 12.x+openssl-1.1.0 user will get TLS1.2.

Most likely, they won't notice. But should they? Should all Node.js 12.x users expect that `tls.MAX_SUPPORTED_VERSION === 'TLSv1.3'`?

I've a couple options:

* I can just cause node to throw when TLS1.3 is attempted, and its not supported. This will happen right now, for example, with node.js 11.x. The problem with this is that TLS1.3 is the default, so most TLS attempts instantly fail (and 204 node test cases fail :-(). It can be worked around by setting `--tls-max-v1.2`, but I suspect the distros that are linking against openssl-1-1.o would not be happy that node doesn't run without env configuration :-(.

* I can cause node to set the default to the max supported by the OpenSSL library in use. This works OK, except that any code the assumption that `tls.MAX_SUPPORTED_VERSION === 'TLSv1.3'` for 12.x, and some users could silently start to negotiate lower protocol versions than expected. Arguably, if you really care, you should explicitly set the TLS max version, possibly with `--tls-max-v1.3`, which would set the max to 1.3 *even if its not supported by the underlying library*, so the app would get an invalid TLS protocol version error, and know there is a problem. But, I don't like the "silently downgrade to a less secure protocol" aspect of this, it seems like a bad idea to me.

* I could just unilaterally make it so that for Node.js 12.x and greater, it only supports openssl 1.1.1 and up. I would prefer this choice. For Node.js 10.x and 11.x, the backport will set the max TLS version to TLS1.2, so the only way to get TLS1.3 is to explicitly enable it. If node.js is running against an underlying openssl library that doesn't support TLS1.3, and the user explicitly requests TLS1.3, they will get an error, which is good and expected.

Thoughts, @nodejs/crypto ?

jasnell commented:
I Think the third option is likely the better choice.

BridgeAR commented:
I removed the backport requested label as it has the opposite effect of what it probably was intended for: if set, this PR will not be automatically pulled in by our tools when backporting PRs (if possible) and instead we'll wait for a manual backport.

sam-github commented:
To elaborate on the third option, which @jasnell and I like: for 11.x and 10.x, it will *still* be possible to link against openssl-1.1.0, even after TLSv1.3 is released for those lines. Node.js won't actually be able to do TLSv1.3 if linked against openssl 1.1.0, but that's OK. Node won't try to do it by default, and if the user explicitly requests TLSv1.3, and node can't do it because it was linked aginst a system openssl that is too old, it will throw an error.

I'll fixup the PR to take this approach, but I'll also have to figure out how this is supposed to work in the build. One of the builds is explicitly to check a compile/link against openssl 1.1.0, and that's no longer valid for all PRs. @nodejs/build any pointers to how to explain this situation to Jenkins?

mhdawson commented:
For 
>But, I don't like the "silently downgrade to a less secure protocol" aspect of this, it seems like a bad idea to me.

I don't see that as a big problem as there are already several cases where you might get TLS 1.2 even if your client supports 1.3 (connecting to server that only supports 1.2 as a good example). As mentioned I think if you really care if you get 1.3 or not then you should request it.

I'm fine with option 3 as well, would be good to get some input from one of the distros as consumers who might link against a different OpenSSL library. @danbev any concern about option 3?

danbev commented:
> @danbev any concern about option 3?

Option 3 works for us, so I'm +1 on that.

sam-github commented:
I went with option 3, then. I'll try to figure out how to tell Jenkins not to try to build/run against OpenSSL-1.1.0.

https://github.com/nodejs/node/pull/26327 landed, so I rebased out my floating patch.

richardlau commented:
> 
> 
> I went with option 3, then. I'll try to figure out how to tell Jenkins not to try to build/run against OpenSSL-1.1.0.
> 

I think the following lines in the version selector script are relevant (currently `/sharedlibs_openssl110/` is excluded for versions < 9): https://github.com/nodejs/build/blob/master/jenkins/scripts/VersionSelectorScript.groovy#L59-L62

sam-github commented:
Thanks, @richardlau.

nodejs/build PR: https://github.com/nodejs/build/pull/1711

sam-github commented:
Some legitimate falures when missing crypto were fixed.

ci: https://ci.nodejs.org/job/node-test-pull-request/21274/

sam-github commented:
Unrelated failure in parallel/test-child-process-pipe-dataflow

Openssl 1.1.0 is expected until nodejs/build#1711 lands.

re-ci: https://ci.nodejs.org/job/node-test-pull-request/21311/

sam-github commented:
ci green, needs review: @nodejs/crypto @nodejs/tsc

addaleax commented on src/tls_wrap.cc:
```diff
@@ -271,8 +280,23 @@ void TLSWrap::EncOut() {
+        // tick.
+        env()->SetImmediate([](Environment* env, void* data) {
+            static_cast<TLSWrap*>(data)->InvokeQueued(0);
+            }, this, object());
```
Tiny style thing:

```suggestion
        }, this, object());
```

addaleax commented on src/tls_wrap.cc:
```diff
@@ -271,8 +280,23 @@ void TLSWrap::EncOut() {
+      if (!in_dowrite_) {
+        InvokeQueued(0);
+      } else {
+        // XXX(@sam-github) If in_dowrite_ is true, appdata was passed to
```
You can add me to the TODO as well if you want, I know I haven’t been the most helpful in investigating this but I’d also still like to address this (/rewrite a lot of this code) at some point

addaleax commented on test/parallel/test-tls-client-reject.js:
```diff
@@ -35,6 +35,8 @@ const options = {
   socket.pipe(socket);
+  // Pipe already ends... but leaving this here tests .end() after .end().
+  // XXX(sam) is that good, or is it bad?
   socket.on('end', () => socket.end());
```
I think it’s just odd, tbh. If we have stream/net tests covering this we can drop it, imo.

sam-github replied:
hard to weed through all the tests to see if its covered, I'll leave it.

addaleax commented on test/parallel/test-tls-min-max-version.js:
```diff
@@ -13,6 +13,7 @@ const DEFAULT_MAX_VERSION = tls.DEFAULT_MAX_VERSION;
 
 function test(cmin, cmax, cprot, smin, smax, sprot, proto, cerr, serr) {
   assert(proto || cerr || serr, 'test missing any expectations');
+  const where = (new Error()).stack.split('\n')[2].replace(/[^(]*/, '');
```
Could you add a comment with an example of what the contents of this variable might end up being? For example, `// e.g. (test/parallel/test-tls-min-max-version.js:1234)`?

sam-github replied:
ok

sam-github commented:
Added changes requested by @addaleax.

ci: https://ci.nodejs.org/job/node-test-pull-request/21316/

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/21338/

https://github.com/nodejs/build/pull/1711#pullrequestreview-212130313 landed, this build should be green.

vsemozhetbyt commented on doc/api/tls.md:
```diff
@@ -1612,16 +1673,16 @@ where `secureSocket` has the same API as `pair.cleartext`.
 [RFC 5077]: https://tools.ietf.org/html/rfc5077
 [RFC 5929]: https://tools.ietf.org/html/rfc5929
-[SSL_METHODS]: https://www.openssl.org/docs/man1.1.0/ssl/ssl.html#Dealing-with-Protocol-Methods
+[SSL_METHODS]: https://www.openssl.org/docs/man1.1.1/ssl/ssl.html#Dealing-with-Protocol-Methods
```
404 Page Not Found.

sam-github commented:
@vsemozhetbyt Thanks, applied your suggestions.

ci: https://ci.nodejs.org/job/node-test-pull-request/21357

BridgeAR commented:
@sam-github this needs a rebase

sam-github commented:
rebased.

ci: https://ci.nodejs.org/job/node-test-pull-request/21436/

rvagg commented on doc/api/tls.md:
```diff
@@ -136,6 +139,8 @@ threshold is exceeded. The limits are configurable:
 The default renegotiation limits should not be modified without a full
 understanding of the implications and risks.
 
+TLSv1.3 does not support renegotiation.
```
for consistency I think this should be prefixed with `Note:`

addaleax replied:
Fwiw, we have (and in particular @trott) has been working towards removing `Note:` from the docs where it’s not really adding any more information.

sam-github replied:
I've also been seeing notes removed, @vsemozhetbyt @Trott , what is current preferences?

vsemozhetbyt replied:
`Note:` style seems to be removed from the style guide in https://github.com/nodejs/node/pull/23903

sam-github replied:
I'm happy to emphasize this some other way, or not, whatever the current style is.

Trott replied:
Having it as a short single-sentence paragraph by itself, the way it is now, is my preference. Setting it apart that way gives it emphasis, but isn't unduly distracting.

For things that truly must be in-your-face emphasized, bold text seems OK, but I'd recommend being very stingy about using it. Typography emphasis gets out of hand very quickly. And we've currently got a mixture of bold, italics, even all caps in a few places... 😞 

Maybe if we do the Google Summer of Docs thing (or whatever it's called), coming up with a consistent emphasis style and giving clear guidance as to what-conditions/how-often it should be used could be a good thing.

Trott replied:
If the main concern is consistency rather than emphasis, I'd suggest removing the other two instances of `Note:` in `tls.md` rather than adding more instances.

Lastly, none of these things I'm writing should be considered blocking. These are my opinions only and the potential for bike shedding is enormous. Just get this PR done and there will sooner or later be a PR to clean up little style things like that across docs (and, where applicable, add to the style guide and the doc-linting rules).

rvagg commented on doc/api/tls.md:
```diff
@@ -176,6 +181,10 @@ as for resumption with session tickets. For debugging, if
 [`tls.TLSSocket.getTLSTicket()`][] returns a value, the session data contains a
 ticket, otherwise it contains client-side session state.
 
+With TLSv1.3, be aware that multiple tickets may be sent by the server,
```
`Note:` again _I think_

rvagg commented on doc/api/tls.md:
```diff
@@ -289,7 +313,18 @@ Old clients that rely on insecure and deprecated RC4 or DES-based ciphers
+- `'TLS_AES_128_CCM_SHA256'`
+- `'TLS_AES_128_CCM_8_SHA256'`
+
+The first 3 are enabled by default. The last 2 `CCM`-based suites are supported
```
Are we following a pattern set by someone else on this? it's an interesting exclusion. one of the main reasons CHACHA20_POLY1305_SHA256 is in there is because it might be more performant on systems that don't have hardware-enabled AES but it'd be slower on systems without it. Are we making choices here that should be left up to users & clients?

sam-github replied:
This exactly duplicates what OpenSSL does, so I think its a safe choice. Unless I misread, these three are also what Chrome and Firefox support: https://www.ssllabs.com/ssltest/viewClient.html?name=Chrome&version=57&platform=Win%207&key=141 and  https://www.ssllabs.com/ssltest/viewClient.html?name=Firefox&version=62&platform=Win%207&key=151 .

rvagg replied:
mm, you're right
```c
#  define TLS_DEFAULT_CIPHERSUITES "TLS_AES_256_GCM_SHA384:" \
                                   "TLS_CHACHA20_POLY1305_SHA256:" \
                                   "TLS_AES_128_GCM_SHA256"
```

rvagg commented:
heroic effort getting us to this point @sam-github, thanks for the hard work!

I've left a few comments, nothing blocking if you feel strongly enough.

I also _really_ don't like the proliferation of meaningless `XXX` in the codebase, perhaps you could find a different way to leave notes. `TODO` is a bit more obvious and leaves hints for people looking to find ways to contribute in the future.

sam-github commented on doc/api/cli.md:
```diff
@@ -443,21 +443,43 @@ added: v4.0.0
+Set default [`maxVersion`][] to `'TLSv1.2'`. Use to disable support for TLSv1.3.
 
-### `--tls-v1.1`
+### `--tls-max-v1.3`
```
TLS1.3 is semver-major, so when its backported to 11.x/10.x, the max will be TLS1.2. To enable TLS1.3, we'll need a CLI option. I could introduce the CLI option just in the backport, but then people who are calling `node --tls-max-1.3` with 10.x and upgrade to node 12.x will start getting 'unrecognized CLI option' errors.

rvagg replied:
that's a good rationale, 👍

sam-github commented on lib/_tls_wrap.js:
```diff
@@ -233,8 +246,8 @@ function onnewsession(sessionId, session) {
 
-  // XXX(sam) no server to emit the event on, but handshake won't continue
-  // unless newSessionDone() is called, should it be?
+  // XXX(@sam-github) no server to emit the event on, but handshake won't
```
its unrelated to TLS1.3, and doesn't "XXX === TODO"? This struck me as potentially problematic during review of existing code, I'm not sure that the situation is reachable with public APIs. No one has noticed it, it completely blocks handshakes if it happens, so it isn't a high priority bug (assuming its a bug at all). I plan to take a look sometime.

sam-github replied:
changed all XXX to TODO, or just comments

sam-github commented:
re: XXX vs TODO vs ????: I don't mind moving this into my local notes, but they might die there, I've too many. I think its useful to leave in-source because if its actually a bug, and it occurs in the wild, someone could see this while trouble-shooting. If there is another format for leaving in-source notes, I'll happily use it, I thought this is how its done in node's source.

sam-github commented:
TLS1.3 (and the node unit tests) don't pass with openssl1.1.1a, it needs updating to openssl1.1.1b (at least) in the docker image, @nodejs/build @refack 

@nodejs/crypto I'm tempted to add a compile time assertion in src/tls_wrap.cc so compilation fails with openssl < 1.1.1b, so that broken node cannot be built. Though, if a person only uses TLS1.2 and below, openssl1.1.1a does work. Do we care? It seems allowing people to link against external openssl libs that don't fully work for node could be seen as either a blessing or a curse for whoever ends up using that custom node build.

@nodejs/build Also, something is wrong with the build scripts, https://ci.nodejs.org/job/node-test-commit-linux-containered/nodes=ubuntu1604_sharedlibs_openssl111_x64/11278/consoleText, at the very end, tap2junit failed.

richardlau commented:
> 
> 
> TLS1.3 (and the node unit tests) don't pass with openssl1.1.1a, it needs updating to openssl1.1.1b (at least) in the docker image, @nodejs/build @refack

Opened https://github.com/nodejs/build/pull/1720.

rvagg commented:
I don't think you need special accounting for 1.1.1a @sam-github, I'll try and get that docker stuff sorted out today or tomorrow so we're doing 1.1.1b in our infra, it's just not an easy process.

rvagg commented:
Oh, and re "XXX", I just don't know what that's supposed to mean, I guess it's a convention that some use but there's so far not much of it in our codebase. If it's supposed to be a note then just leave a comment, if it's a TODO then make it a TODO so it's obvious to future contributors looking for something to help with. Anyone else got an opinion on this?

sam-github commented:
I learned XXX from the node code base, but just counted, TODO is indeed more popular. Changed all XXX in this to TODO.

rvagg commented:
all looks good to me, anything left before getting this landed?

addaleax commented:
@sam-github Could you rebase this?

sam-github commented:
Just FYI, rebase is in progress, but the V8 update on master needs recompilation. It'll be a while until I have a build.

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/21646/

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/21648/

sam-github commented:
ci: https://ci.nodejs.org/job/node-test-pull-request/21672/

sam-github commented:
Since it was rebased, `git node land` wants another approval before its landable.

sam-github commented:
Landed in 42dbaed

Yay! Thanks for all the help.

gauravmahto commented:
Will this be backported to v10.x?

sam-github commented:
I took a shot at it, https://github.com/sam-github/node/commits/tls1.3-v10.x, but it has so many deps that weren't backported it just exploded in difficulty, and I gave up. It isn't a priority for me, not now at least, and you are literally the only person who I have ever heard express any interest in TLS1.3 on 10.x...

Maybe someone else who feels the drive will have to step and do the backport. @gauravmahto , perhaps you?

gauravmahto commented:
@sam-github I would love to. But I'll definitely need the guidance related to where to start and through the process. Last time I tried building on my machine (Windows), it kinda worked but several tests started to fail. Maybe this is the perfect time for me to give it another shot. :)

gauravmahto commented:
@sam-github I built and ran the tests. Looks like everything worked just fine. :)
Still, I would need proper guidance. looking forward to it.

sam-github commented:
I don't have a lot to offer outside of https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#the-process-of-making-changes and https://github.com/nodejs/node/blob/master/COLLABORATOR_GUIDE.md#how-are-lts-branches-managed

Feel free to PR a branch containing whatever you think needs doing to v10.x-staging. Make sure it tests locally, and once it passes on travis, ask someone to start a full CI and see if it passes.
